[BUG]: Adding field in Fieldbook from BMS using BrAPI: No Categories for categorical traits.

### Description

When connecting BMS and Fieldbook App for loading Fields (Studies in BMS) there is an issue for the traits that are Categorical. There valid categories are no recive in the Fieldbook App.

Steps to reproduce it:

1. Create a study in BMS. Add Traits in Observations tab, be sure at least one trait is Categorical.

![image](https://github.com/PhenoApps/Field-Book/assets/78973920/f6d107a5-6eff-4e97-a376-c92bd98ec3d5)


2. Connect KSU Fieldbook to BMS using BrAPI. (Using

3. In the App, go to Fields and add a Study using BrAPI

4. Go to Traits and open the Categorical Trait to see the categories. Or go to Collect and move to the Categorical trait.

Video of the steps: https://drive.google.com/file/d/12HN7LiOFA1-Y66XHpsDif-v4f_ppKbqH/view?usp=sharing



Actual result
KSU Fieldbook DO NOT display the Categories for the categorical trait.
![image](https://github.com/PhenoApps/Field-Book/assets/78973920/da89ee22-58d2-4cdd-b18f-f808b913d954)
![image](https://github.com/PhenoApps/Field-Book/assets/78973920/3e6a3fdd-71b9-49e0-b5ae-29f23fa8acb9)

Video of the steps:



Note: Tested in old versions of Fieldbook app (5.2.0) against BMS v24 and the categories are displayed properly in the app. In Version 5.3 and 5.4 (Beta) is not working properly.

### Expected behavior

KSU Fieldbook display the Categories for the categorical trait. Categories are send to the App in the correct order.
![image](https://github.com/PhenoApps/Field-Book/assets/78973920/a89ce4b6-abf3-461f-ad10-a996b5fdfef8)


### Field Book version

5.3.0 and 5.4.0

> @jpietragalla @abatac Please test out the [generated apk](https://github.com/PhenoApps/Field-Book/pull/647#issuecomment-1611851325) to see if this issue persists.

We tested the apk but it has issues with BrAPI Field Import. When the Preview Field button is clicked, the application crashes.

I looked at the logs and I see these errors:

```
2023-06-29 10:49:21.211 16444-16556/? E/BrAPIServiceV1: Error fetching observation levels
    io.swagger.client.ApiException: 
        at io.swagger.client.ApiClient.handleResponse(ApiClient.java:924)
        at io.swagger.client.ApiClient$1.onResponse(ApiClient.java:879)
        at com.squareup.okhttp.Call$AsyncCall.execute(Call.java:177)
        at com.squareup.okhttp.internal.NamedRunnable.run(NamedRunnable.java:33)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
        at java.lang.Thread.run(Thread.java:1012)
```

```
2023-06-29 10:42:17.314 15715-15715/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.fieldbook.tracker, PID: 15715
    java.lang.IllegalStateException: Could not execute method for android:onClick
        at androidx.appcompat.app.AppCompatViewInflater$DeclaredOnClickListener.onClick(AppCompatViewInflater.java:446)
        at android.view.View.performClick(View.java:7506)
        at android.view.View.performClickInternal(View.java:7483)
        at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0)
        at android.view.View$PerformClick.run(View.java:29335)
        at android.os.Handler.handleCallback(Handler.java:942)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7929)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at androidx.appcompat.app.AppCompatViewInflater$DeclaredOnClickListener.onClick(AppCompatViewInflater.java:441)
        at android.view.View.performClick(View.java:7506) 
        at android.view.View.performClickInternal(View.java:7483) 
        at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0) 
        at android.view.View$PerformClick.run(View.java:29335) 
        at android.os.Handler.handleCallback(Handler.java:942) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loopOnce(Looper.java:201) 
        at android.os.Looper.loop(Looper.java:288) 
        at android.app.ActivityThread.main(ActivityThread.java:7929) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936) 
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.fieldbook.tracker.brapi.model.BrapiObservationLevel.getObservationLevelName()' on a null object reference
        at com.fieldbook.tracker.brapi.service.BrAPIServiceV1.getPlotDetails(BrAPIServiceV1.java:511)
        at com.fieldbook.tracker.brapi.BrapiLoadDialog.buildStudyDetails(BrapiLoadDialog.java:158)
        at com.fieldbook.tracker.brapi.BrapiLoadDialog.onStart(BrapiLoadDialog.java:94)
        at android.app.Dialog.show(Dialog.java:333)
        at com.fieldbook.tracker.activities.brapi.BrapiActivity.saveStudy(BrapiActivity.java:242)
        at com.fieldbook.tracker.activities.brapi.BrapiActivity.buttonClicked(BrapiActivity.java:226)
        at java.lang.reflect.Method.invoke(Native Method) 
        at androidx.appcompat.app.AppCompatViewInflater$DeclaredOnClickListener.onClick(AppCompatViewInflater.java:441) 
        at android.view.View.performClick(View.java:7506) 
        at android.view.View.performClickInternal(View.java:7483) 
        at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0) 
        at android.view.View$PerformClick.run(View.java:29335) 
        at android.os.Handler.handleCallback(Handler.java:942) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loopOnce(Looper.java:201) 
        at android.os.Looper.loop(Looper.java:288) 
        at android.app.ActivityThread.main(ActivityThread.java:7929) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936) 
```

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -157,6 +157,7 @@ dependencies {
     implementation 'com.github.breeding-insight:brapi:2.0.3'
     implementation("com.squareup.okhttp3:okhttp:4.9.3")
     implementation("com.squareup.okhttp3:logging-interceptor:4.9.3")
+    implementation("io.gsonfire:gson-fire:1.8.4") //required v1
 
     implementation 'org.jsoup:jsoup:1.8.1'
     implementation 'net.sourceforge.jexcelapi:jxl:2.6.10'

--- a/app/src/main/java/com/fieldbook/tracker/brapi/service/BrAPIServiceV1.java
+++ b/app/src/main/java/com/fieldbook/tracker/brapi/service/BrAPIServiceV1.java
@@ -441,6 +441,23 @@ public class BrAPIServiceV1 extends AbstractBrAPIService implements BrAPIService
                                final Function<BrapiStudyDetails, Void> function,
                                final Function<Integer, Void> failFunction) {
         try {
+
+            //level is not a required field, so passing null here is fine
+            //otherwise this was causing a NPE when observationLevel was null
+            //check if observationLevel is null
+            @Nullable
+            String levelName = null;
+            if (observationLevel != null) {
+
+                if (observationLevel.getObservationLevelName() != null) {
+
+                    levelName = observationLevel.getObservationLevelName();
+
+                }
+            }
+
+            final String level = levelName;
+
             final AtomicInteger currentPage = new AtomicInteger(0);
             final Integer pageSize = Integer.parseInt(context.getSharedPreferences(GeneralKeys.SHARED_PREF_FILE_NAME, 0)
                     .getString(GeneralKeys.BRAPI_PAGE_SIZE, "50"));
@@ -481,13 +498,6 @@ public class BrAPIServiceV1 extends AbstractBrAPIService implements BrAPIService
                         }else {
                             try {
 
-                                //level is not a required field, so passing null here is fine
-                                //otherwise this was causing a NPE when observationLevel was null
-                                @Nullable String level = null;
-                                if (observationLevel != null) {
-                                    level = observationLevel.getObservationLevelName();
-                                }
-
                                 studiesApi.studiesStudyDbIdObservationunitsGetAsync(
                                         studyDbId, level, currentPage.get(), pageSize,
                                         getBrapiToken(), this);
@@ -508,7 +518,7 @@ public class BrAPIServiceV1 extends AbstractBrAPIService implements BrAPIService
             };
 
             studiesApi.studiesStudyDbIdObservationunitsGetAsync(
-                    studyDbId, observationLevel.getObservationLevelName(), 0, pageSize,
+                    studyDbId, level, 0, pageSize,
                     getBrapiToken(), callback);
 
         } catch (ApiException e) {

--- a/app/src/main/java/com/fieldbook/tracker/brapi/service/BrAPIServiceV1.java
+++ b/app/src/main/java/com/fieldbook/tracker/brapi/service/BrAPIServiceV1.java
@@ -1006,6 +1006,7 @@ public class BrAPIServiceV1 extends AbstractBrAPIService implements BrAPIService
                     c.setValue(value);
                     c.setLabel(value);
                 }
+                scale.add(c);
             }
             return CategoryJsonUtil.Companion.encode(scale);
         } catch (Exception e) {

